src/Test.UnitTests.Sarif/Writers/SarifLoggerTests.cs

```diff
@@ -975,7 +1013,10 @@ private void TestForLoggingOption(LogFilePersistenceOptions loggingOption)
                 SarifLogger logger;
 
                 // Validates overload that accept a path argument.
-                using (logger = new SarifLogger(fileName, loggingOption))
+                using (logger = new SarifLogger(fileName, 
+                                                loggingOption,
+                                                levels: new List<FailureLevel> { FailureLevel.Warning, FailureLevel.Error },
```

Comment:
>levels: new List<FailureLevel> { FailureLevel.Warning, FailureLevel.Error }, [](start = 48, length = 76)

I would have expected these args to be defaults on the SarifLogger class, so you wouldn't have had to make them explicit in the tests. To provide default enumerables, check for null in the relevant ctor and initialize a new List instance on detecting null.

Reply:
I thought about putting default values but the validation in BaseLogger is the *only* validation that is run.  That is, BaseLogger is the one place we've defined rules for what command line options can be input.  BaseLogger, and by extension, SarifLogger isn't asking for its caller to "tell me what you think that I need to know", it's asking for the true values of the options passed so it can accurately validate them.  If we put default values in SarifLogger, we are correcting bad input.  And if we're going to correct bad input, why are we throwing exceptions at all?  Why not just overwrite their options if they ever pass options that don't make sense?

---
In reply to: [563215266](https://github.com/microsoft/sarif-sdk/pull/2262#discussion_r563215266) [](ancestors = 563215266)